### PR TITLE
HDFS-16667. Use malloc for buffer allocation in uriparser2

### PR DESCRIPTION
**Pull request by GauthamBanasandra**
### Description of PR
Currently, a variable is used to specify the array size in uriparser2 -
https://github.com/apache/hadoop/blob/34e548cb62ed21c5bba7a82f5f1489ca6bdfb8c4/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfspp/third_party/uriparser2/uriparser2/uriparser2.c#L71-L79

This results in the following error on Windows -
```
H:\hadoop-hdfs-project\hadoop-hdfs-native-client\src\out\build\x64-Debug\cl : command line warning D9025: overriding '/W4' with '/w' 
    uriparser2.c
H:\hadoop-hdfs-project\hadoop-hdfs-native-client\src\main\native\libhdfspp\third_party\uriparser2\uriparser2\uriparser2.c(74,23): error C2057: expected constant expression 
H:\hadoop-hdfs-project\hadoop-hdfs-native-client\src\main\native\libhdfspp\third_party\uriparser2\uriparser2\uriparser2.c(74,23): error C2466: cannot allocate an array of constant size 0 
H:\hadoop-hdfs-project\hadoop-hdfs-native-client\src\main\native\libhdfspp\third_party\uriparser2\uriparser2\uriparser2.c(74,24): error C2133: 'buffer': unknown size 
```

I've used malloc to fix this.

### How was this patch tested?
1. Tested this locally by building on my Windows system.
2. Hadoop Jenkins CI validation.

### For code changes:

- [x] Does the title or this PR starts with the corresponding JIRA issue id (e.g. 'HADOOP-17799. Your PR title ...')?
- [ ] Object storage: have the integration tests been executed and the endpoint declared according to the connector-specific documentation?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE`, `LICENSE-binary`, `NOTICE-binary` files?

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |  45m 32s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  40m 39s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   3m 57s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 38s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  67m 14s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 19s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   3m 56s |  |  the patch passed  |
| +1 :green_heart: |  cc  |   3m 56s |  |  the patch passed  |
| +1 :green_heart: |  golang  |   3m 56s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   3m 56s |  |  the patch passed  |
| -1 :x: |  blanks  |   0m  0s | [/blanks-tabs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/1/artifact/out/blanks-tabs.txt) |  The patch 4 line(s) with tabs.  |
| +1 :green_heart: |  mvnsite  |   0m 22s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  22m 10s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  96m 43s |  |  hadoop-hdfs-native-client in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 45s |  |  The patch does not generate ASF License warnings.  |
|  |   | 239m 39s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/4576 |
| Optional Tests | dupname asflicense compile cc mvnsite javac unit codespell detsecrets golang |
| uname | Linux b457491c3273 4.15.0-175-generic #184-Ubuntu SMP Thu Mar 24 17:48:36 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 4f43c44c5769365eb34c3f40c877bdaef5ee1e27 |
| Default Java | Red Hat, Inc.-1.8.0_332-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/1/testReport/ |
| Max. process+thread count | 636 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs-native-client U: hadoop-hdfs-project/hadoop-hdfs-native-client |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/1/console |
| versions | git=2.9.5 maven=3.6.3 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfspp/third_party/uriparser2/uriparser2/uriparser2.c [hunk 1]
```diff
@@ -71,9 +71,11 @@ static const char *copy_path(const UriPathSegmentA *ps, char **buffer) {
 static int parse_int(const char *first, const char *after_last) {
 	const int size = after_last - first;
 	if (size) {
-		char buffer[size + 1];
+		char* buffer = (char*) malloc(size + 1);
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfspp/third_party/uriparser2/uriparser2/uriparser2.c]: Isn't the compiler being too pedantic? The old code should be safe.

Comment (GauthamBanasandra) [reply hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfspp/third_party/uriparser2/uriparser2/uriparser2.c]: Local arrays are stored on the stack. The stack has a fixed limit. Thus, MSVC tries to play safe by making the size a compile-time requirement, forcing us to use the heap. I don't know how GCC solves this though.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 51s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  85m 25s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   3m 58s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 38s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  | 112m 10s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 20s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   3m 52s |  |  the patch passed  |
| +1 :green_heart: |  cc  |   3m 52s |  |  the patch passed  |
| +1 :green_heart: |  golang  |   3m 52s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   3m 52s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |   0m 22s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  22m 16s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  88m 30s |  |  hadoop-hdfs-native-client in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 51s |  |  The patch does not generate ASF License warnings.  |
|  |   | 231m 50s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/4576 |
| Optional Tests | dupname asflicense compile cc mvnsite javac unit codespell detsecrets golang |
| uname | Linux c9777e842920 4.15.0-175-generic #184-Ubuntu SMP Thu Mar 24 17:48:36 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / df2323ca40fb4579c6be56bee9b8a780c20243d5 |
| Default Java | Red Hat, Inc.-1.8.0_332-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/3/testReport/ |
| Max. process+thread count | 615 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs-native-client U: hadoop-hdfs-project/hadoop-hdfs-native-client |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/3/console |
| versions | git=2.9.5 maven=3.6.3 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |  29m 12s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  27m 29s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   4m 54s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 50s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  58m 16s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 26s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   4m 12s |  |  the patch passed  |
| +1 :green_heart: |  cc  |   4m 12s |  |  the patch passed  |
| +1 :green_heart: |  golang  |   4m 12s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   4m 12s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |   0m 27s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  23m 42s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 107m 12s |  |  hadoop-hdfs-native-client in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 49s |  |  The patch does not generate ASF License warnings.  |
|  |   | 227m  4s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/4576 |
| Optional Tests | dupname asflicense compile cc mvnsite javac unit codespell detsecrets golang |
| uname | Linux a10d05b9b421 4.15.0-175-generic #184-Ubuntu SMP Thu Mar 24 17:48:36 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / df2323ca40fb4579c6be56bee9b8a780c20243d5 |
| Default Java | Red Hat, Inc.-1.8.0_312-b07 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/3/testReport/ |
| Max. process+thread count | 623 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs-native-client U: hadoop-hdfs-project/hadoop-hdfs-native-client |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/3/console |
| versions | git=2.27.0 maven=3.6.3 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |  27m 27s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  33m 54s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   4m  5s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 43s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  71m 25s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 22s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   3m 56s |  |  the patch passed  |
| +1 :green_heart: |  cc  |   3m 56s |  |  the patch passed  |
| +1 :green_heart: |  golang  |   3m 56s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   3m 56s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |   0m 27s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  33m 43s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 113m 38s |  |  hadoop-hdfs-native-client in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 51s |  |  The patch does not generate ASF License warnings.  |
|  |   | 255m  0s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/4576 |
| Optional Tests | dupname asflicense compile cc mvnsite javac unit codespell detsecrets golang |
| uname | Linux 7eb9a1960edb 4.15.0-175-generic #184-Ubuntu SMP Thu Mar 24 17:48:36 UTC 2022 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / df2323ca40fb4579c6be56bee9b8a780c20243d5 |
| Default Java | Debian-11.0.15+10-post-Debian-1deb10u1 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/3/testReport/ |
| modules | C: hadoop-hdfs-project/hadoop-hdfs-native-client U: hadoop-hdfs-project/hadoop-hdfs-native-client |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/3/console |
| versions | git=2.20.1 maven=3.6.0 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   1m 23s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  28m 20s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   5m  9s |  |  trunk passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  compile  |   4m 57s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  mvnsite  |   0m 45s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  64m 44s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 22s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   4m 41s |  |  the patch passed with JDK Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1  |
| +1 :green_heart: |  cc  |   4m 41s |  |  the patch passed  |
| +1 :green_heart: |  golang  |   4m 41s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   4m 41s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   4m 41s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  cc  |   4m 41s |  |  the patch passed  |
| +1 :green_heart: |  golang  |   4m 41s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   4m 41s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |   0m 24s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  25m 32s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  97m  6s |  |  hadoop-hdfs-native-client in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 40s |  |  The patch does not generate ASF License warnings.  |
|  |   | 202m 58s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/4576 |
| Optional Tests | dupname asflicense compile cc mvnsite javac unit codespell detsecrets golang |
| uname | Linux 232d09e8622b 4.15.0-175-generic #184-Ubuntu SMP Thu Mar 24 17:48:36 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / df2323ca40fb4579c6be56bee9b8a780c20243d5 |
| Default Java | Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Private Build-11.0.15+10-Ubuntu-0ubuntu0.20.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/3/testReport/ |
| Max. process+thread count | 606 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs-native-client U: hadoop-hdfs-project/hadoop-hdfs-native-client |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-4576/3/console |
| versions | git=2.25.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Review by goiri — APPROVED** [review]

**Pull request merged**